utils/scarb-ui/src/components/typed.rs

```diff
@@ -43,8 +51,28 @@ impl<'a> TypedMessage<'a> {
             message,
             type_style: None,
             skip_type_for_text: true,
+            type_suffix_for_text: None,
+            extra: Default::default(),
+        }
+    }
+
+    /// Apply additional type suffix.
+    ///
+    /// The suffix will only be shown in the text mode.
+    pub fn with_type_suffix_for_text(self, suffix: &'a str) -> Self {
+        Self {
+            type_suffix_for_text: Some(suffix),
+            ..self
         }
     }
+
+    /// Add additional key-value pair to the message.
+    ///
+    /// This will be exposed in the structured output.
+    pub fn with_extra(mut self, key: &'a str, value: &'a str) -> Self {
+        self.extra.insert(key, value);
+        self
+    }
```

Comment:
I'm afraid this API might be too powerful. 😕 It enables pushing information just for the human display, or just for the JSON-one, which we strive to avoid (not at all cost, though) Have you considered just storing the `error_code: &str` in `TypedMessage` directly?